### PR TITLE
Miscellaneous Storage Performance Improvements

### PR DESCRIPTION
**Pull request by alzimmermsft**
# Description

Fixes: https://github.com/Azure/azure-sdk-for-java/issues/28220
Fixes: https://github.com/Azure/azure-sdk-for-java/issues/24555

- Replaces simple `Pattern`s with `indexOf` and substring. `Pattern` has much higher overhead and in hot paths can be a major slowdown.
- Used `List` and `Map` size constructors when size is known ahead of time.
- Removed `HashMap` creation when existing `Map` was null by returning null.
- Replaced usage of `JacksonAdapter` with `HttpHeader`-based constructor which is many times faster with less overhead.

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/implementation/StorageImplUtils.java [hunk 1]
```diff
@@ -370,8 +347,12 @@ public static String convertStorageExceptionMessage(String message, HttpResponse
                 && response.getHeaders().getValue(ERROR_CODE) != null) {
                 // Use a constant compiled Pattern as the match pattern is always the same and String.replaceFirst
                 // will compile the match String into a Pattern internally on each call.
```
Comment (ibrahimrabab) [inline sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/implementation/StorageImplUtils.java:349]: nit: we could get rid of these comments (not necessary to complete this PR though)

**Review by ibrahimrabab — APPROVED** [review]

**Comment by alzimmermsft** [comment]
Overrode one failing live test due to dependencies failing to download, all others passed.

**Pull request merged**